src/core/qgsvectorlayerutils.cpp

```diff
@@ -420,7 +413,8 @@ QgsFeature QgsVectorLayerUtils::createFeature( const QgsVectorLayer *layer, cons
 
     // 4. provider side default literal
     // note - deliberately not using else if!
-    if ( !v.isValid() && fields.fieldOrigin( idx ) == QgsFields::OriginProvider )
+    if ( ( !v.isValid() || ( fields.at( idx ).constraints().constraints() & QgsFieldConstraints::ConstraintUnique  && QgsVectorLayerUtils::valueExists( layer, idx, v ) ) )
+         && fields.fieldOrigin( idx ) == QgsFields::OriginProvider )
```

Comment:
@signedav 

is there any difference between this check and the one on line 402?
Point 4 will be always executed if point 3 is true.

In any event, it make no sense to me to check for unique in case of default value clauses, unless maybe when they are evaluated immediately.

So, what happens here is on postgis pk fields with a sequence is that for the first digitized feature you get it correctly, but for the second one, this check will be true (because we have already added a feature with the default clause value) and an invalid value will be assigned instead of the nextval, breaking the NOT NULL constraint.

Apparently this is not an issue on gpkg, but that's a separate issue.

Reply:
I didn't see completely through yet. But just to say already: The check 4 is the same as 3, but in case there has been set a unique value to `v` in 3, check 4 is not true anymore because of `QgsVectorLayerUtils::valueExists( layer, idx, v )`.

Reply:
As I understand you and the bug reports it's about that it checks in `QgsVectorLayerUtils::valueExists( layer, idx, v )` for the existence of the "nextval(xy)"-value, correct?

Means there should be something like ` && v != layer->dataProvider()->defaultValueClause( providerIndex )` in check 4, but I have to recheck to be secure...

Are you already on a fix for that, or should I care for it?

Reply:
Thanks for taking care of this, I'm leaving it to you with the recommendation that this function is extremely important: it is used by processing in-place algorithms and in many other core functions.

Please write tests for this case.

Let me know if I can help.

Reply:
Thanks for preliminary clarification @elpaso 
I opened a PR for that (quite small) fix #8510
Could you have a look?

Reply:
@signedav sure, I'll give it a try ASAP!